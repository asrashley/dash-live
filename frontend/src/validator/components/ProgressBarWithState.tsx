import { useComputed } from "@preact/signals";

import { ProgressBar } from "./ProgressBar";
import {
  UseValidatorWebsocketHook,
  ValidatorState,
} from "../hooks/useValidatorWebsocket";

const stateClassMap = {
  [ValidatorState.IDLE]: "text-bg-secondary",
  [ValidatorState.ACTIVE]: "bg-success-subtle text-dark",
  [ValidatorState.CANCELLING]: "bg-danger",
  [ValidatorState.CANCELLED]: "bg-warning-subtle text-dark",
  [ValidatorState.DONE]: "bg-success",
};

export interface ProgressBarWithStateProps {
  progress: UseValidatorWebsocketHook["progress"];
  state: UseValidatorWebsocketHook["state"];
}
export function ProgressBarWithState({
  progress,
  state,
}: ProgressBarWithStateProps) {
  const stateClass = useComputed<string>(
    () =>
      `position-absolute top-0 start-50 translate-middle badge rounded-pill ${
        stateClassMap[state.value]
      }`
  );

  return (
    <div className="position-relative">
      <div className="card progress">
        <ProgressBar progress={progress} />
      </div>
      <div className={stateClass} data-testid="validator-state-badge">{state}</div>
    </div>
  );
}
