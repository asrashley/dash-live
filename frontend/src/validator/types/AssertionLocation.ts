export interface AssertionLocation {
    filename: string;
    line: number;
}
