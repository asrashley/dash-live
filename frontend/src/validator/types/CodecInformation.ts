export interface CodecDetails {
    label: string;
    error?: string;
    details: string[];
}

export interface CodecInformation {
    codec: string;
    error?: string;
    details: CodecDetails[];
}
