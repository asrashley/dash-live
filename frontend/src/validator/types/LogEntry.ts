export type LogEntry = {
    id: number;
    level: 'error' | 'info' | 'debug';
    text: string;
};
