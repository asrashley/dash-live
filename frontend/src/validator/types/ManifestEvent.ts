
export interface ManifestEvent {
    text: string[];
}