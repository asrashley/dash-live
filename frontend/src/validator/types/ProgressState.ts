export interface ProgressState {
    currentValue?: number;
    minValue: number;
    maxValue: number;
    text: string;
    finished?: boolean;
    error?: boolean;
 }
