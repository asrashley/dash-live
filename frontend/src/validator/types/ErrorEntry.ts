import { AssertionLocation } from "./AssertionLocation";

export interface ErrorEntry {
    assertion: AssertionLocation;
    location: [number, number];
    clause?: string;
    msg: string;
}