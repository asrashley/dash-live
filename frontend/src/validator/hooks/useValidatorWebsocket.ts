import { useCallback, useEffect, useMemo, useRef } from 'preact/hooks';
import { batch, type ReadonlySignal, useComputed, useSignal } from '@preact/signals';
import { io } from 'socket.io-client';
import { decode } from 'codec-string';

import { LogEntry } from '../types/LogEntry';
import { ProgressState } from '../types/ProgressState';
import { CodecInformation } from '../types/CodecInformation';
import { ErrorEntry } from '../types/ErrorEntry';
import { ManifestLine } from '../types/ManifestLine';
import { ManifestEvent } from '../types/ManifestEvent';
import { ValidatorSettings } from '../types/ValidatorSettings';

export interface ValidatorProgressEvent {
    pct: number;
    text: string;
    aborted?: boolean;
    finished?: boolean;
}

export interface ValidatorFinishedEvent {
    startTime: number;
    endTime: number;
    aborted: boolean;
}

export interface InstallStreamCommand {
    filename: string;
    title: string;
    prefix: string;
}

export enum ValidatorState {
    IDLE = 'idle',
    ACTIVE = 'active',
    CANCELLING = 'cancelling',
    CANCELLED = 'cancelled',
    DONE = 'done',
}

export interface UseValidatorWebsocketHook {
    codecs: ReadonlySignal<CodecInformation[]>;
    errors: ReadonlySignal<ErrorEntry[]>;
    log: ReadonlySignal<LogEntry[]>;
    manifest: ReadonlySignal<ManifestLine[]>;
    progress: ReadonlySignal<ProgressState>;
    state: ReadonlySignal<ValidatorState>;
    result: ReadonlySignal<ValidatorFinishedEvent|undefined>;
    start: (settings: ValidatorSettings) => void;
    cancel: () => void;
}

export function useValidatorWebsocket(wssUrl: string): UseValidatorWebsocketHook {
    const socket = useMemo(() => io(wssUrl), [wssUrl]);
    const log = useSignal<LogEntry[]>([]);
    const progress = useSignal<ProgressState>({ minValue: 0, maxValue: 100, finished: false, error: false, text: ''});
    const manifestText = useSignal<string[]>([]);
    const codecs = useSignal<CodecInformation[]>([]);
    const errors = useSignal<ErrorEntry[]>([]);
    const state = useSignal<ValidatorState>(ValidatorState.IDLE);
    const result = useSignal<ValidatorFinishedEvent|undefined>();
    const manifest = useComputed<ManifestLine[]>(() => {
        const lines: ManifestLine[] = manifestText.value.map((text: string, idx: number) => ({
            text,
            line: idx + 1,
            hasError: false,
            errors: [],
        }));
        for (const err of errors.value) {
            const { location, clause, msg } = err;
            const [ start, end ] = location;
            const text = clause ? `${clause}: ${msg}` : msg;
            for(let i=start; i <= end; ++i) {
                lines[i - 1].hasError = true;
            }
            lines[start - 1].errors.push(text);
        }
        return lines;
    });
    const nextMsgId = useRef<number>(1);

    const addLogMessage = useCallback((msg: Omit<LogEntry, 'id'>) => {
        log.value = [
            ...log.value,
            {
                ...msg,
                id: nextMsgId.current,
            },
        ];
        nextMsgId.current += 1;
    }, [log]);

    const onProgress = useCallback((data: ValidatorProgressEvent) => {
        const { aborted, pct, text, finished } = data;
        batch(() => {
            progress.value = {
                ...progress.value,
                text,
                finished,
                error: aborted,
                currentValue: Math.round(10 * pct) / 10,
            };
        });
        if (finished) {
           socket.emit('cmd', { method: 'done' });
        }
    }, [progress, socket]);

    const onManifest = useCallback(({ text }: ManifestEvent) => {
        manifestText.value = text;
    }, [manifestText]);

    const onCodecs = useCallback((codecNames: string[]) => {
        const data: CodecInformation[] = codecNames.map((codec: string) => {
            const { error, decodes } = decode(codec);
            const ci: CodecInformation = {
                codec,
                error,
                details: decodes.map(({parsed, error, label}) => ({
                    label,
                    error,
                    details: parsed.map(p => p.decode),
                })),
            };
            return ci;
        });
        codecs.value = data;
    }, [codecs]);

    const onManifestErrors = useCallback((data: ErrorEntry[]) => {
        errors.value = data;
    }, [errors]);

    const onFinished = useCallback((data: ValidatorFinishedEvent) => {
        batch(() => {
            if (state.value == ValidatorState.CANCELLING) {
                state.value = ValidatorState.CANCELLED;
            } else if (state.value == ValidatorState.ACTIVE) {
                state.value = ValidatorState.DONE;
            }
            progress.value = {
                ...progress.value,
                finished: true,
            };
            result.value = data;
        });
    }, [progress, result, state]);

    const onInstallStream = useCallback(({ filename, title, prefix }: InstallStreamCommand) => {
        addLogMessage({
            level: 'info',
            text: `Installing new stream from ${filename}`,
        });
        socket.emit('cmd', {
            method: 'save',
            filename,
            prefix,
            title,
        });
    }, [addLogMessage, socket]);

    const start = useCallback((settings: ValidatorSettings) => {
        batch(() => {
            state.value = ValidatorState.ACTIVE;
            manifestText.value = [];
            log.value = [];
            errors.value = [];
            progress.value = {
                minValue: 0, maxValue: 100, finished: false, error: false, text: ''
            };
            result.value = undefined;
        });
        socket.emit('cmd', {
            method: 'validate',
            ...settings,
        });
    }, [errors, log, manifestText, progress, result, socket, state]);

    const cancel = useCallback(() => {
        batch(() => {
            state.value = ValidatorState.CANCELLING;
            progress.value = {
                ...progress.value,
                error: true,
            };
        });
        socket.emit('cmd', {
            method: 'cancel',
        });
    }, [progress, socket, state]);

    useEffect(() => {
        socket.on('codecs', onCodecs);
        socket.on('finished', onFinished);
        socket.on('install', onInstallStream);
        socket.on('log', addLogMessage);
        socket.on('manifest', onManifest);
        socket.on('manifest-errors', onManifestErrors);
        socket.on('progress', onProgress);

        return () => {
            socket.off('codecs', onCodecs);
            socket.off('finished', onFinished);
            socket.off('install', onInstallStream);
            socket.off('log', addLogMessage);
            socket.off('manifest', onManifest);
            socket.off('manifest-errors', onManifestErrors);
            socket.off('progress', onProgress);
        };
    }, [addLogMessage, onCodecs, onFinished, onManifest, onManifestErrors, onProgress, onInstallStream, socket]);

    return { codecs, errors, log, manifest, progress, result, state, start, cancel };
}
