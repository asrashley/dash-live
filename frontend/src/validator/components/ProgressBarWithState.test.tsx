import { beforeEach, describe, expect, test } from "vitest";
import { signal } from "@preact/signals";

import { ValidatorState } from "../hooks/useValidatorWebsocket";
import { renderWithProviders } from "../../test/renderWithProviders";
import { ProgressBarWithState } from "./ProgressBarWithState";
import { ProgressState } from "../types/ProgressState";

describe("ProgressBarWithState component", () => {
  const progress = signal<ProgressState>({
    minValue: 0,
    maxValue: 100,
    finished: false,
    error: false,
    text: "",
  });
  const state = signal<ValidatorState>(ValidatorState.DISCONNECTED);

  beforeEach(() => {
    state.value = ValidatorState.DISCONNECTED;
    progress.value = {
      minValue: 0,
      maxValue: 100,
      finished: false,
      error: false,
      text: "",
    };
  });

  test.each<ValidatorState>([
    ValidatorState.DISCONNECTED,
    ValidatorState.IDLE,
    ValidatorState.ACTIVE,
    ValidatorState.CANCELLING,
    ValidatorState.CANCELLED,
    ValidatorState.DONE,
  ])("renders when state=%s", (valState: ValidatorState) => {
    state.value = valState;
    progress.value.text = `renders when state=${valState}`;
    const { asFragment, getByTestId, getByText } = renderWithProviders(
      <ProgressBarWithState state={state} progress={progress} />
    );
    const badge = getByTestId("validator-state-badge") as HTMLElement;
    expect(badge.innerHTML).toEqual(valState);
    getByText(progress.value.text);
    expect(asFragment()).toMatchSnapshot();
  });
});
