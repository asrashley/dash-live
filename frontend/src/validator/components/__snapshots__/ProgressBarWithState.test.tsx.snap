// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressBarWithState component > renders when state=active 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <div
      class="card progress"
    >
      <div>
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          class="progress-bar progress-bar-striped visually-hidden"
          role="progressbar"
        >
          NaN%
        </div>
        <div
          class="progress-text"
        >
          renders when state=active
        </div>
      </div>
    </div>
    <div
      class="position-absolute top-0 start-50 translate-middle badge rounded-pill bg-success-subtle text-dark"
      data-testid="validator-state-badge"
    >
      active
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBarWithState component > renders when state=cancelled 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <div
      class="card progress"
    >
      <div>
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          class="progress-bar progress-bar-striped visually-hidden"
          role="progressbar"
        >
          NaN%
        </div>
        <div
          class="progress-text"
        >
          renders when state=cancelled
        </div>
      </div>
    </div>
    <div
      class="position-absolute top-0 start-50 translate-middle badge rounded-pill bg-warning-subtle text-dark"
      data-testid="validator-state-badge"
    >
      cancelled
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBarWithState component > renders when state=cancelling 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <div
      class="card progress"
    >
      <div>
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          class="progress-bar progress-bar-striped visually-hidden"
          role="progressbar"
        >
          NaN%
        </div>
        <div
          class="progress-text"
        >
          renders when state=cancelling
        </div>
      </div>
    </div>
    <div
      class="position-absolute top-0 start-50 translate-middle badge rounded-pill bg-danger"
      data-testid="validator-state-badge"
    >
      cancelling
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBarWithState component > renders when state=disconnected 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <div
      class="card progress"
    >
      <div>
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          class="progress-bar progress-bar-striped visually-hidden"
          role="progressbar"
        >
          NaN%
        </div>
        <div
          class="progress-text"
        >
          renders when state=disconnected
        </div>
      </div>
    </div>
    <div
      class="position-absolute top-0 start-50 translate-middle badge rounded-pill text-bg-dark"
      data-testid="validator-state-badge"
    >
      disconnected
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBarWithState component > renders when state=done 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <div
      class="card progress"
    >
      <div>
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          class="progress-bar progress-bar-striped visually-hidden"
          role="progressbar"
        >
          NaN%
        </div>
        <div
          class="progress-text"
        >
          renders when state=done
        </div>
      </div>
    </div>
    <div
      class="position-absolute top-0 start-50 translate-middle badge rounded-pill bg-success"
      data-testid="validator-state-badge"
    >
      done
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBarWithState component > renders when state=idle 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <div
      class="card progress"
    >
      <div>
        <div
          aria-valuemax="100"
          aria-valuemin="0"
          class="progress-bar progress-bar-striped visually-hidden"
          role="progressbar"
        >
          NaN%
        </div>
        <div
          class="progress-text"
        >
          renders when state=idle
        </div>
      </div>
    </div>
    <div
      class="position-absolute top-0 start-50 translate-middle badge rounded-pill text-bg-secondary"
      data-testid="validator-state-badge"
    >
      idle
    </div>
  </div>
</DocumentFragment>
`;
